muxers/mplex/src/io.rs

```diff
@@ -0,0 +1,771 @@
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use bytes::Bytes;
+use crate::{MplexConfig, MaxBufferBehaviour};
+use crate::codec::{Codec, Frame, LocalStreamId, RemoteStreamId};
+use log::{debug, trace};
+use fnv::FnvHashMap;
+use futures::{prelude::*, ready, stream::Fuse};
+use futures::task::{ArcWake, waker_ref, WakerRef};
+use futures_codec::Framed;
+use parking_lot::Mutex;
+use std::collections::{VecDeque, hash_map::Entry};
+use std::{cmp, io, mem, sync::Arc, task::{Context, Poll, Waker}};
+
+pub use std::io::{Result, Error, ErrorKind};
+
+/// A multiplexed I/O stream.
+pub struct Multiplexed<C> {
+    /// The current operating status.
+    status: Status,
+    /// The underlying I/O stream.
+    io: Fuse<Framed<C, Codec>>,
+    /// The configuration.
+    config: MplexConfig,
+    /// Buffer of received frames that have not yet been consumed.
+    buffer: Vec<Frame<RemoteStreamId>>,
+    /// Whether a flush is pending before reading frames can proceed.
+    pending_flush: bool,
+    /// Pending frames to send at the next opportunity.
+    ///
+    /// An opportunity for sending pending frames is every flush
+    /// or read operation. In the former case, sending of all
+    /// pending frames must complete before the flush can complete.
+    /// In the latter case, the read operation can proceed even
+    /// if some or all of the pending frames cannot be sent.
+    pending_frames: VecDeque<Frame<LocalStreamId>>,
+    /// The substreams that are considered at least half-open.
+    open_substreams: FnvHashMap<LocalStreamId, SubstreamState>,
+    /// The ID for the next outbound substream.
+    next_outbound_stream_id: LocalStreamId,
+    /// Registry of wakers for pending tasks interested in reading.
+    notifier_read: Arc<NotifierRead>,
+    /// Registry of wakers for pending tasks interested in writing.
+    notifier_write: Arc<NotifierWrite>,
+    /// Registry of wakers for pending tasks interested in opening
+    /// an outbound substream, when the configured limit is reached.
+    notifier_open: Arc<NotifierOpen>,
+}
+
+/// The operation status of a `Multiplexed` I/O stream.
+#[derive(Debug)]
+enum Status {
+    /// The stream is considered open and healthy.
+    Ok,
```

Comment:
Wouldn't `Open` be more descriptive, given that there is a `Closed`? Not a strong opinion.

Reply:
Done in https://github.com/libp2p/rust-libp2p/pull/1769/commits/adca14ffba41da1cf77b692c59790694bcef12e6.